sw/device/lib/dif/dif_uart.c

```diff
@@ -0,0 +1,392 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/dif/dif_uart.h"
+
+#include <stddef.h>
+
+#include "sw/device/lib/dif/register.h"
+#include "uart_regs.h"  // Generated.
+
+// Fake base address
+#define UART0_BASE_ADDR 0x0
+
+#define UART_RX_FIFO_SIZE_BYTES 32u
+#define UART_TX_FIFO_SIZE_BYTES 32u
+
+// Validate the UART ON/OFF value
+static bool uart_on_off_valid(dif_uart_on_off_t val) {
+  return (val == dif_uart_enable || val == dif_uart_disable);
+}
+
+// Check the status register for TX FIFO full
+static bool uart_tx_full(const dif_uart_t *uart) {
+  return reg32_get_bit(uart->base_addr, UART_STATUS(0), UART_STATUS_TXFULL);
+}
+
+// Check the status register for RX FIFO empty
+static bool uart_rx_empty(const dif_uart_t *uart) {
+  return reg32_get_bit(uart->base_addr, UART_STATUS(0), UART_STATUS_RXEMPTY);
+}
+
+// Return number of bytes in RX FIFO
+static uint8_t uart_rx_fifo_bytes(const dif_uart_t *uart) {
```

Comment:
Why does this return a byte rather than a `size_t`?

Reply:
The idea was to return the smallest possible type, as the FIFO is up to 32 bytes deep, but I agree, size_t is more appropriate.

Reply:
done